hail/src/main/c/Region2.cpp

```diff
@@ -0,0 +1,89 @@
+#include "hail/Region2.h"
+#include "hail/Upcalls.h"
+#include <memory>
+#include <vector>
+#include <utility>
+#include "catch.hpp"
+
+namespace hail {
+
+RegionPool::Region::Region(RegionPool * pool) :
+pool_(pool),
+block_offset_(0),
+current_block_(pool_->get_block()),
+used_blocks_(std::vector<char *>()),
+big_chunks_(std::vector<char *>()),
+parents_(std::vector<std::shared_ptr<Region2>>()) { }
+
+char * Region2::allocate_new_block() {
+  used_blocks_.push_back(current_block_);
+  current_block_ = pool_->get_block();
+  block_offset_ = 0;
+  return current_block_;
+}
+
+char * Region2::allocate_big_chunk(ssize_t n) {
+  char * ptr = (char *) malloc(n);
+  big_chunks_.push_back(ptr);
+  return ptr;
+}
+
+void Region2::clear() {
+  block_offset_ = 0;
+  for (char * block : used_blocks_) { pool_->free_blocks_.push_back(block); }
+  for (char * chunk : big_chunks_) { free(chunk); }
+  used_blocks_.clear();
+  big_chunks_.clear();
+  parents_.clear();
+}
+
+std::shared_ptr<Region2> Region2::get_region() {
+  return pool_->get_region();
+}
+
+void Region2::add_reference_to(std::shared_ptr<Region2> region) {
+  parents_.push_back(region);
```

Comment:
This is copying the shared_ptr twice, then deleting one copy: it copies once into the `region` parameter, then again into `parents_`, then `region` is deleted at the end of the scope. Since refcount modifications have to be atomic (something we might think about fixing by rolling our own reference-counting smart pointer), it's best to minimize them. I would make this `parents_.push_back(std::move(region))`, avoiding the copy, while keeping the signature of `add_reference_to` that explicitly says a copy of the shared_ptr is being made.